### PR TITLE
GAEA support, authoritative branches, non-nested global grids

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
### Switch to Authoritative Branches

This switches from HAFS code to the authoritative branches of:
- tracker
- ufs-weather-model
- UFS_UTILS
- UPP (via ufs-weather-model)

### Disable Unused Components

Also, it disables compilation of components that aren't used:
- GSI
- hycom_utils
- ww3_utils

### Ports

The HAFS-AR runs on GAEA and Hera except graphics. Pygraf graphics are only installed on Hera.

### Uniform and Stretched Global Options

The stretched global and uniform global work. That's a global FV3 with no nest. Access these with:

grid.gtype=uniform or stretch
grid.nest_grids=0

### Workarounds

The UPP crashes on GAEA, so we're using the inline post instead.

There's a new CCPP suite FV3_global_nest_v0 which is the FV3_global_nest_v1 with the old gravity wave drag. This is a temporary measure for script testing. The workflow will be updated soon to use the right suite.

The sorc/customizations contains patches that the sorc/build_*.sh scripts apply before compiling. Here are the reasons for them:

1. UFS_UTILS = GAEA support
2. tracker = GAEA support
3. ufs-weather-model = FV3_global_nest_v0 suite (temporary; will be removed soon)

## Issues addressed (optional)
No issue submitted. Problem is listed in description.

## Dependencies (optional)
None.

## Contributors (optional)
None.

## Tests conducted
Ran an atmospheric river case and verified the output looked like an atmosphere. Did this on GAEA and Hera.

## Application-level regression test status
None.

### CODE REVIEW

File: modulefiles/hafs.hera.lua [hunk 1]
```diff
@@ -6,122 +6,65 @@ prepend_path("MODULEPATH", "/contrib/sutils/modulefiles")
...
+  end
+end
 
-cdo_ver=os.getenv("cdo_ver") or "1.9.8"
-load(pathJoin("cdo", cdo_ver))
 
-rocoto_ver=os.getenv("rocoto_ver") or "1.3.3"
-load(pathJoin("rocoto", rocoto_ver))
+prepend_path("MODULEPATH", "/scratch1/NCEPDEV/hwrf/noscrub/local/modulefiles")
+load(pathJoin("python","wcoss2_env"))
```
Comment (guoqing-noaa) [inline modulefiles/hafs.hera.lua:67]: load "python/wcoss2_env" on Hera?

File: parm/WaterInTheWestC384NoNest.conf [hunk 2]
```diff
@@ -0,0 +1,176 @@
+# This is a UNIX conf file that contains ALL information relating to
+# the HAFS basin-focused global-nesting configuration.
```
Comment (guoqing-noaa) [inline parm/WaterInTheWestC384NoNest.conf]: line 2 comment says "global-nesting" but this is NoNest.conf

File: parm/WaterInTheWestC384Uniform.conf [hunk 3]
```diff
@@ -0,0 +1,176 @@
+# This is a UNIX conf file that contains ALL information relating to
+# the HAFS basin-focused global-nesting configuration.
```
Comment (guoqing-noaa) [inline parm/WaterInTheWestC384Uniform.conf]: line 2 comment says "global-nesting" but this is Uniform.conf

File: parm/forecast/globnest/input.nml.nonest.tmp
Comment (guoqing-noaa) [inline parm/forecast/globnest/input.nml.nonest.tmp:1]: Do we need to track "tmp" files?

File: sorc/hafs_tools.fd/sorc/hafs_obs_preproc/CMakeLists.txt [hunk 4]
```diff
@@ -87,7 +87,7 @@ target_link_options(${exe_name} PRIVATE
 
 target_link_libraries(
   ${exe_name}
-  bufr::bufr_d
+  bufr::bufr_4
```
Comment (guoqing-noaa) [inline sorc/hafs_tools.fd/sorc/hafs_obs_preproc/CMakeLists.txt:90]: This is always curious to me and I did not get time to check this. Are we sure bufr_d is exactly the same as bufr_4?

File: sorc/customizations/hafs_tracker.fd/gaea-c5-patch.diff
Comment (guoqing-noaa) [inline sorc/customizations/hafs_tracker.fd/gaea-c5-patch.diff:1]: I assume including "patch.diff" files is a PR requirement?

**Review by guoqing-noaa — COMMENTED** [review]
@SamuelTrahanNOAA Thanks for inviting me to review this PR. Great work! I was not able to test this on Gaea as I haven't set up a Gaea account yet but I visually went through the changes and it looks good to me. I asked a few questions but they may be general questions and not specific to this PR.

File: parm/forecast/globnest/input.nml.nonest.tmp
Comment (SamuelTrahanNOAA) [reply parm/forecast/globnest/input.nml.nonest.tmp:1]: tmp = template

File: sorc/hafs_tools.fd/sorc/hafs_obs_preproc/CMakeLists.txt (hunk 4 above)
Comment (SamuelTrahanNOAA) [reply sorc/hafs_tools.fd/sorc/hafs_obs_preproc/CMakeLists.txt:90]: No. If memory serves:
bufr_d = 8 byte integers, 4 byte reals
bufr_4 = 4 byte integers, 4 byte reals
bufr_8 = 8 byte integers, 8 byte reals

The latest versions of bufr only know how to build bufr_4

File: sorc/customizations/hafs_tracker.fd/gaea-c5-patch.diff
Comment (SamuelTrahanNOAA) [reply sorc/customizations/hafs_tracker.fd/gaea-c5-patch.diff:1]: Yes. This is an unfortunate workaround for the lack of GAEA C5 support in the components.

I'll do a PR to update the tracker soon.

File: modulefiles/hafs.hera.lua (hunk 1 above)
Comment (SamuelTrahanNOAA) [reply modulefiles/hafs.hera.lua:67]: That's the Python that the HAFS developers use. I don't know why they give it that name.

File: parm/WaterInTheWestC384NoNest.conf (hunk 2 above)
Comment (SamuelTrahanNOAA) [reply parm/WaterInTheWestC384NoNest.conf]: Now it says this:

```bash
# This is a UNIX conf file that contains ALL information relating to
# the HAFS stretched globe configuration with no nest.
```

File: parm/WaterInTheWestC384Uniform.conf (hunk 3 above)
Comment (SamuelTrahanNOAA) [reply parm/WaterInTheWestC384Uniform.conf]: Now it says this:

```bash
# This is a UNIX conf file that contains ALL information relating to
# the HAFS uniform C384 global configuration (with no nest)
```

File: sorc/hafs_tools.fd/sorc/hafs_obs_preproc/CMakeLists.txt (hunk 4 above)
Comment (guoqing-noaa) [reply sorc/hafs_tools.fd/sorc/hafs_obs_preproc/CMakeLists.txt:90]: Thanks, @SamuelTrahanNOAA ! (I think you mean the "integers" and "reals" to be swapped)

I took a little bit of time to dig into this change. It looks like they changed to use a Fortran/C interface to consolidate _4, _d and _8 versions into one library (but renamed as _4, which is unexpected to me).
https://github.com/NOAA-EMC/NCEPLIBS-bufr/issues/78

**Review by guoqing-noaa — APPROVED** [review]
LGTM. Thanks!

File: parm/WaterInTheWestC384NoNest.conf [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+istart_nest=25
+jstart_nest=73
+iend_nest=744
+jend_nest=648
+regional_esg=no
+
+[forecast]
+ccpp_suite_regional=FV3_global_nest_v0
+ccpp_suite_glob=FV3_global_nest_v0
+ccpp_suite_nest=FV3_global_nest_v0
```
Comment (AndersJensen-NOAA) [inline parm/WaterInTheWestC384NoNest.conf]: Is this needed in the no nest config?

File: parm/WaterInTheWestC384Uniform.conf
Comment (AndersJensen-NOAA) [inline parm/WaterInTheWestC384Uniform.conf:1]: Do we need all of these variables defined in 2 separate config files for Uniform and NoNest -- it seems like there is a lot of overlap here, particularly the physics definitions and [config].

File: rocoto/WaterInTheWest.sh [hunk 6]
```diff
@@ -10,9 +10,13 @@ EXPT=$(basename ${HOMEhafs})
 SUBEXPT=$EXPT
 conf=../parm/WaterInTheWest.conf
 opts="-t -f"
+
+# To enable graphics, uncomment this. But only on Hera!
```
Comment (AndersJensen-NOAA) [inline rocoto/WaterInTheWest.sh]: Can there be a check for this ... `if hera`

File: sorc/link_fix.sh [hunk 7]
```diff
@@ -21,6 +21,8 @@ elif [ ${target} == "orion" ]; then
   FIXROOT=/work/noaa/hwrf/noscrub/hafs-fix-files/hafs-${FIXversion}-fix/fix
 elif [ ${target} == "jet" ]; then
   FIXROOT=/lfs4/HFIP/hwrf-data/hafs-fix-files/hafs-${FIXversion}-fix/fix
+elif [[ $target =~ gaea ]]; then
+  FIXROOT=/lustre/f2/dev/Samuel.Trahan/hafs/hafs-fix-files/hafs-${FIXversion}-fix/fix
```
Comment (AndersJensen-NOAA) [inline sorc/link_fix.sh:25]: Is there a plan to move this out of your personal directory?

**Review by AndersJensen-NOAA — APPROVED** [review]

File: parm/WaterInTheWestC384NoNest.conf (hunk 5 above)
Comment (SamuelTrahanNOAA) [reply parm/WaterInTheWestC384NoNest.conf]: I haven't tried removing it.

File: parm/WaterInTheWestC384Uniform.conf
Comment (SamuelTrahanNOAA) [reply parm/WaterInTheWestC384Uniform.conf:1]: No. I could modify the files to override only the necessary variables. You would do this in WaterInTheWest.sh:

./run_hafs.py ... other arguments ... ../parm/WaterInTheWestC384.conf ../parm/WaterInTheWestC384NoNest.conf

or, for a uniform global,

./run_hafs.py ... other arguments ... ../parm/WaterInTheWestC384.conf ../parm/WaterInTheWestC384Uniform.conf

File: rocoto/WaterInTheWest.sh (hunk 6 above)
Comment (SamuelTrahanNOAA) [reply rocoto/WaterInTheWest.sh]: That script doesn't know if it is on Hera. With a bit of work, I could have run_hafs.py override the run_pygraf if you're on GAEA.

File: sorc/link_fix.sh (hunk 7 above)
Comment (SamuelTrahanNOAA) [reply sorc/link_fix.sh:25]: Do we have somewhere else to put it?

File: parm/WaterInTheWestC384NoNest.conf (hunk 5 above)
Comment (SamuelTrahanNOAA) [reply parm/WaterInTheWestC384NoNest.conf]: If I reduce this file to only overridden values, those lines will be gone.

File: sorc/link_fix.sh (hunk 7 above)
Comment (mark-a-potts) [reply sorc/link_fix.sh:25]: We should be able to put it in the role.epic space somewhere.

Comment (SamuelTrahanNOAA) [reply sorc/link_fix.sh:25]: @mark-a-potts -

Please copy this directory to a communal area:

/lustre/f2/dev/Samuel.Trahan/hafs/hafs-fix-files

and let me know when it is there. I'll update the link_fix.sh.

File: parm/WaterInTheWestC384Uniform.conf
Comment (SamuelTrahanNOAA) [reply parm/WaterInTheWestC384Uniform.conf:1]: I'm testing this change now on Hera.

Comment (SamuelTrahanNOAA) [reply parm/WaterInTheWestC384Uniform.conf:1]: It works. On GAEA, pygraf is disabled even if you set config.run_pygraf=yes

File: rocoto/WaterInTheWest.sh (hunk 6 above)
Comment (SamuelTrahanNOAA) [reply rocoto/WaterInTheWest.sh]: It works. On GAEA, pygraf is disabled even if you set config.run_pygraf=yes

File: parm/WaterInTheWestC384Uniform.conf
Comment (SamuelTrahanNOAA) [reply parm/WaterInTheWestC384Uniform.conf:1]: They now have only the values they must override from WaterInTheWestC384.conf

File: sorc/link_fix.sh (hunk 7 above)
Comment (SamuelTrahanNOAA) [reply sorc/link_fix.sh:25]: I don't want to wait any longer. I'll merge this now. If you provide me a directory later, I'll update the script.

**Pull request merged**